### ISSUE TITLE
Tween.transform throws exception with null safety

### ISSUE THREAD

**Issue by fourman-alex** [flutter/flutter#79558]
## Steps to Reproduce

when `begin=null` and `tween.transform(0)` or `end=null` and `tween.transform(1.0)`

### To reproduce run the following code with null safety enabled
```
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: Home(),
    );
  }
}

class Home extends StatelessWidget {
  final tween = BorderRadiusTween(begin: null, end: null);
  @override
  Widget build(BuildContext context) {
    tween.transform(0);
    return Scaffold();
  }
}
```

### The culprit is the `as T` call which obviously fails for null. 

> ```
>   @override
>   T transform(double t) {
>     if (t == 0.0)
>       return begin as T;
>     if (t == 1.0)
>       return end as T;
>     return lerp(t);
>   }
> 
>```

**Comment by TahaTesser** [comment]
Hi @fourman-alex 
<details> 
<summary>logs</summary> 

```bash
[   +6 ms] Flutter run key commands.
[   +3 ms] r Hot reload. 🔥🔥🔥
[   +2 ms] R Hot restart.
[        ] h Repeat this help message.
[        ] d Detach (terminate "flutter run" but leave application running).
[        ] c Clear the screen
[        ] q Quit (terminate the application on the device).
[        ] 💪 Running with sound null safety 💪
[   +1 ms] An Observatory debugger and profiler on iPhone 12 Pro Max is available at: http://127.0.0.1:54791/hFk3CIbyQtE=/
[ +265 ms] 
                    ══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
                    The following _CastError was thrown building Home(dirty):
                    type 'Null' is not a subtype of type 'BorderRadius' in type cast
                    
                    The relevant error-causing widget was:
                      Home file:///Users/nevercode/AndroidStudioProjects/master_flutter/lib/main.dart:12:13
                    
                    When the exception was thrown, this was the stack:
                    #0      Tween.transform (package:flutter/src/animation/tween.dart:322:20)
                    #1      Home.build (package:master_flutter/main.dart:21:11)
                    #2      StatelessElement.build (package:flutter/src/widgets/framework.dart:4648:28)
                    #3      ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:4574:15)
                    #4      Element.rebuild (package:flutter/src/widgets/framework.dart:4267:5)
                    #5      ComponentElement._firstBuild (package:flutter/src/widgets/framework.dart:4553:5)
                    #6      ComponentElement.mount (package:flutter/src/widgets/framework.dart:4548:5)
                    ...     Normal element mounting (24 frames)
                    #30     Element.inflateWidget (package:flutter/src/widgets/framework.dart:3611:14)
                    #31     MultiChildRenderObjectElement.inflateWidget (package:flutter/src/widgets/framework.dart:6221:36)
                    #32     MultiChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:6232:32)
                    ...     Normal element mounting (170 frames)
                    #202    Element.inflateWidget (package:flutter/src/widgets/framework.dart:3611:14)
                    #203    MultiChildRenderObjectElement.inflateWidget (package:flutter/src/widgets/framework.dart:6221:36)
                    #204    MultiChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:6232:32)
                    ...     Normal element mounting (360 frames)
                    #564    Element.inflateWidget (package:flutter/src/widgets/framework.dart:3611:14)
                    #565    Element.updateChild (package:flutter/src/widgets/framework.dart:3363:18)
                    #566    RenderObjectToWidgetElement._rebuild (package:flutter/src/widgets/binding.dart:1189:16)
                    #567    RenderObjectToWidgetElement.mount (package:flutter/src/widgets/binding.dart:1159:5)
                    #568    RenderObjectToWidgetAdapter.attachToRenderTree.<anonymous closure> (package:flutter/src/widgets/binding.dart:1104:18)
                    #569    BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2535:19)
                    #570    RenderObjectToWidgetAdapter.attachToRenderTree (package:flutter/src/widgets/binding.dart:1103:13)
                    #571    WidgetsBinding.attachRootWidget (package:flutter/src/widgets/binding.dart:937:7)
                    #572    WidgetsBinding.scheduleAttachRootWidget.<anonymous closure> (package:flutter/src/widgets/binding.dart:917:7)
                    (elided 11 frames from class _RawReceivePortImpl, class _Timer, dart:async, and dart:async-patch)
                    
                    ════════════════════════════════════════════════════════════════════════════════════════════════════
[ +
```

</details>

<details> 
<summary>flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel stable, 2.0.4, on macOS 11.2.3 20D91 darwin-arm, locale
    en-GB)
    • Flutter version 2.0.4 at /Users/nevercode/Code/flutter_stable
    • Framework revision b1395592de (4 days ago), 2021-04-01 14:25:01 -0700
    • Engine revision 2dce47073a
    • Dart version 2.12.2

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/nevercode/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • Java binary at: /Users/nevercode/Library/Application
      Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/201.7042882/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for
      more details.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Users/nevercode/Library/Application
      Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/201.7042882/Android
      Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.55.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (3 available)
    • iPhone 12 Pro Max (mobile) • 42B7F3DF-3053-419B-BA8F-FF1218F7A174 • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-14-4 (simulator)
    • macOS (desktop)            • macos                                •
      darwin-arm64   • macOS 11.2.3 20D91 darwin-arm
    • Chrome (web)               • chrome                               •
      web-javascript • Google Chrome 89.0.4389.114

! Doctor found issues in 1 category.
```

```bash
[✓] Flutter (Channel master, 2.1.0-13.0.pre.444, on macOS 11.2.3 20D91 darwin-arm, locale
    en-GB)
    • Flutter version 2.1.0-13.0.pre.444 at /Users/nevercode/Code/flutter_master
    • Framework revision 824b1ad3f8 (2 days ago), 2021-04-03 15:24:02 -0400
    • Engine revision e393800c63
    • Dart version 2.13.0 (build 2.13.0-196.0.dev)

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/nevercode/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • Java binary at: /Users/nevercode/Library/Application
      Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/201.7042882/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for more details.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Users/nevercode/Library/Application
      Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/201.7042882/Android
      Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.55.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (3 available)
    • iPhone 12 Pro Max (mobile) • 42B7F3DF-3053-419B-BA8F-FF1218F7A174 • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-14-4 (simulator)
    • macOS (desktop)            • macos                                • darwin-arm64   •
      macOS 11.2.3 20D91 darwin-arm
    • Chrome (web)               • chrome                               • web-javascript •
      Google Chrome 89.0.4389.114

! Doctor found issues in 1 category.
```
	
</details>

Thanks for filing the issue!

**Comment by goderbauer** [comment]
@xu-baolin I assigned this to you based on the fact that you have a PR out to fix this.
